src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs

```diff
@@ -17,6 +18,8 @@ namespace Microsoft.Extensions.DependencyInjection
     /// </summary>
     public static class ActivatorUtilities
     {
+        private static ConcurrentDictionary<Type, ConstructorInfoEx[]> s_constructorInfoEx = new();
```

Comment:
Do we need to worry about collectible types?

Reply:
Can you provide details on when this occurs? (unloadable Assembly\AppDomain\ASP.NET feature?; .NET Framework makes it easier).

Other caches, [such as this one by System.Text.Json](https://github.com/dotnet/runtime/blob/8a2cc6d6fdf2eeea783af49b1e470fbe2c564ee7/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/PolymorphicTypeResolver.cs#L18) uses a similar cache based on System.Type.

I assume we'd base the cache on a `WeakReferenece` if we did this.

Reply:
It would be best to attach this cache to some sort of DI context instance. It would address the unloadability concern. Is there a type like that in DI today?

Reply:
Wouldn't ConditionalWeakTable work here?

Reply:
We should at least support "ClearCache()" for hot reload, so I added that for now.

> It would be best to attach this cache to some sort of DI context instance. It would address the unloadability concern. Is there a type like that in DI today?

Since DI here is based on IServiceProvider, there isn't anything more but I'll look a bit more.

Reply:
System.Text.Json does have the same issue as I mentioned, at least for caches held by the "default" JsonSerializerOptions which is a static variable.

The reflection caches for at least the IL-based emit generation also have an issue here where they reference types in IL that may be unloaded - although we could set the "owning" module correctly based on DeclaringType.

Reply:
> I am not sure which cache you are talking about. Where is this cache?

Not really a cache I guess, but [a DynamicMethod used in this file](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Reflection/InvokerEmitUtil.cs#L33)  that doesn't set the owning module correctly (AFAIK).

Reply:
Note I pushed a commit that adds unload support using `ConditionalWeakTable`. I'll add tests if this approach is acceptable.

Reply:
> [DynamicMethod used in this file](https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Reflection/InvokerEmitUtil.cs#L33) that doesn't set the owning module correctly (AFAIK).

This module argument is not owning module for the lifetime purposes. The dynamic method lifetime is not tied to the lifetime of the dynamic method.